libbpf-tools/cpudist.bpf.c

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "cpudist.h"
+#include "helpers.h"
+
+#define TASK_RUNNING 0
+
+const volatile bool targ_per_process = false;
+const volatile bool targ_per_thread = false;
+const volatile bool targ_offcpu = false;
+const volatile bool targ_ms = false;
+const volatile pid_t targ_tgid = -1;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 4);
+	__type(key, u32);
+	__type(value, u64);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 4);
+	__type(key, u32);
+	__type(value, u64);
+} global_hist SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 4);
+	__type(key, struct piddata);
+	__type(value, u64);
+} per_task_hist SEC(".maps");
+
+static __always_inline void store_start(u32 tgid, u32 pid, u64 ts)
+{
+	if (targ_tgid != -1 && targ_tgid != tgid)
+		return;
+	bpf_map_update_elem(&start, &pid, &ts, 0);
+}
+
+static __always_inline void update_hist(u32 tgid, u32 pid, u64 ts)
+{
+	u64 delta, *countp, *tsp, count = 1;
+	struct piddata piddata = {};
+	void *histp, *slotp;
+	u32 slot;
+
+	if (targ_tgid != -1 && targ_tgid != tgid)
+		return;
+
+	tsp = bpf_map_lookup_elem(&start, &pid);
+	if (!tsp || ts < *tsp)
+		return;
+
+	delta = ts - *tsp;
+	if (targ_ms)
+		delta /= 1000000;
+	else
+		delta /= 1000;
+	slot = bpf_log2l(delta);
+	if (!targ_per_process && !targ_per_thread) {
+		histp = &global_hist;
+		slotp = &slot;
+	} else {
+		piddata.slot = slot;
+		histp = &per_task_hist;
+		slotp = &piddata;
+		if (targ_per_process) {
+			piddata.pid = tgid;
+		} else if (targ_per_thread) {
+			piddata.pid = pid;
+		}
+	}
```

Comment:
this feels like an overkill, honestly. Why not unify global and per process/thread data structures by encoding global one as special pid==-1 or something like that? it would all be quite simpler already.

But then, instead of having each histogram slot as a separate entry in hashtable, wouldn't it make more sense to have a fixed-sized array for each possible pid and update corresponding slots? Sure, it's a bit more wasteful, but then how many slots do we need per each pid, right? Plus you can make each slot just u32, instead of u64.

With such a change, you probably also don't need entire dict stuff in user-space, which would further minimize overhead.

Reply:
Agree, your design is much better than mine. I complicated things. I've done with such a change, it looks much clearer and less overhead. Thank you.